src/Hazelcast.Net/CP/SessionAwareSemaphore.cs

```diff
@@ -0,0 +1,230 @@
+﻿// Copyright (c) 2008-2023, Hazelcast, Inc. All Rights Reserved.
+// 
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// 
+// http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using System.Threading.Tasks;
+using Hazelcast.Clustering;
+using Hazelcast.Core;
+using Hazelcast.DistributedObjects;
+using Hazelcast.Protocol;
+using Hazelcast.Protocol.Codecs;
+using Hazelcast.Protocol.Models;
+
+namespace Hazelcast.CP;
+
+internal class SessionAwareSemaphore : CPDistributedObjectBase, ISemaphore
+{
+    public const int DrainSessionAcqCount = 1024;
+
+    private readonly CPSessionManager _sessionManager;
+
+    public SessionAwareSemaphore(string name, CPGroupId groupId, Cluster cluster, CPSessionManager sessionManager)
+        : base(ServiceNames.Semaphore, name, groupId, cluster)
+    {
+        _sessionManager = sessionManager;
+    }
+
+    // this semaphore is session-aware, therefore it can use a plain "thread id" in the
+    // client request messages - and for semaphores, this has to be a somewhat not constant
+    // yet not totally random number, which is used internally by the cluster to manage
+    // queues and stuff - but, two simultaneous operations on the same "thread id" can
+    // interrupt each other, so this id has to somehow represent the .NET async flow (cannot
+    // be the plain .NET managed thread) and therefore we use the AsyncContext.
+    private static long GetThreadId() => AsyncContext.Current.Id;
```

Comment:
Because it's a semaphore it's totally unrelated to context. It's purely helping the server manage its internal queues.

See https://hazelcast.slack.com/archives/C01JU7ZJYGP/p1703672606977649